concordium-node/src/consensus_ffi/ffi.rs

```diff
@@ -1665,6 +1796,233 @@ pub fn get_consensus_ptr(
+                invoker_account_address_ptr,
+                invoker_contract_index,
+                invoker_contract_subindex,
+                amount,
+                receive_name_ptr,
+                receive_name_len,
+                parameter_ptr,
+                parameter_len,
+                energy,
+                &mut out_data,
+            )
+        };
+        DryRun::check_result(res, "invoke instance")?;
+        Ok(out_data)
+    }
+
+    /// Set the current block time for the dry-run session.
+    pub fn set_timestamp(
+        &mut self,
+        new_timestamp: crate::grpc2::types::Timestamp,
+    ) -> Result<Vec<u8>, tonic::Status> {
+        let mut out_data: Vec<u8> = Vec::new();
+
+        let res = unsafe { dryRunSetTimestamp(self.handle, new_timestamp.value, &mut out_data) };
+        DryRun::check_result(res, "set timestamp")?;
+        Ok(out_data)
+    }
+
+    /// Mint a specified amount and credit it to the specified account.
+    pub fn mint_to_account(
+        &mut self,
+        mint: crate::grpc2::types::DryRunMintToAccount,
+    ) -> Result<Vec<u8>, tonic::Status> {
+        use crate::grpc2::Require;
+        let mut out_data: Vec<u8> = Vec::new();
+        let account_address =
+            crate::grpc2::types::account_address_to_ffi(mint.account.as_ref().require()?)
+                .require()?;
+
+        let amount = mint.amount.require()?;
+        let res = unsafe {
+            dryRunMintToAccount(self.handle, account_address, amount.value, &mut out_data)
+        };
+        DryRun::check_result(res, "mint to account")?;
+        Ok(out_data)
+    }
+
+    /// Run a transaction in the current dry-run state, updating the state if it
+    /// succeeds.
+    pub fn transaction(
+        &mut self,
+        request: crate::grpc2::types::DryRunTransaction,
+    ) -> Result<Vec<u8>, tonic::Status> {
+        use crate::grpc2::Require;
+        let mut out_data: Vec<u8> = Vec::new();
+        let energy = request.energy_amount.as_ref().require()?.value;
+
+        let sender_address =
+            crate::grpc2::types::account_address_to_ffi(request.sender.as_ref().require()?)
+                .require()?;
+        let encoded_payload =
+            concordium_base::transactions::EncodedPayload::try_from(request.payload.require()?)?;
+        let payload_bytes: Vec<u8> = encoded_payload.into();
+        let signature_count = request.signatures.len();
+        let mut signature_vec: Vec<u8> = Vec::with_capacity(signature_count * 2);
+        for sig in request.signatures {
+            signature_vec.push(
+                sig.credential.try_into().map_err(|_| {
+                    tonic::Status::invalid_argument("Credential index out of bounds")
+                })?,
+            );
+            signature_vec.push(
+                sig.key
+                    .try_into()
+                    .map_err(|_| tonic::Status::invalid_argument("Key index out of bounds"))?,
+            );
+        }
```

Comment:
As I mentioned there, there's no real benefit to using a map. It would have to be a map of maps, which also means introducing a new type for the inner map, and then the target would be empty values. It's just more convenient to work with a list of pairs and not worry too much about duplicates.

Reply:
I disagree but not enough to persist at it.